source/opt/scalar_replacement_pass.cpp

```diff
@@ -0,0 +1,598 @@
+    iter = where.InsertBefore(std::move(newStore));
+    get_def_use_mgr()->AnalyzeInstDefUse(&*iter);
+    context()->set_instr_block(&*iter, block);
+  }
+}
+
+bool ScalarReplacementPass::ReplaceAccessChain(
+    ir::Instruction* chain, const std::vector<ir::Instruction*>& replacements) {
+  // Replaces the access chain with either another access chain (with one fewer
+  // indexes) or a direct use of the replacement variable.
+  uint32_t indexId = chain->GetSingleWordInOperand(1u);
+  const ir::Instruction* index = get_def_use_mgr()->GetDef(indexId);
+  size_t indexValue = GetConstantInteger(index);
+  if (indexValue > replacements.size()) {
+    // Out of bounds access, this is illegal IR.
+    return false;
+  } else {
+    const ir::Instruction* var = replacements[indexValue];
+    if (chain->NumInOperands() > 2) {
+      // Replace input access chain with another access chain.
+      ir::BasicBlock::iterator chainIter(chain);
+      uint32_t replacementId = TakeNextId();
+      std::unique_ptr<ir::Instruction> replacementChain(new ir::Instruction(
+          context(), chain->opcode(), chain->type_id(), replacementId,
+          std::initializer_list<ir::Operand>{
+              {SPV_OPERAND_TYPE_ID, {var->result_id()}}}));
+      // Add the remaining indexes.
+      for (uint32_t i = 2; i < chain->NumInOperands(); ++i) {
+        ir::Operand copy(chain->GetInOperand(i));
+        replacementChain->AddOperand(std::move(copy));
+      }
+      auto iter = chainIter.InsertBefore(std::move(replacementChain));
+      get_def_use_mgr()->AnalyzeInstDefUse(&*iter);
+      context()->set_instr_block(&*iter, context()->get_instr_block(chain));
+      context()->ReplaceAllUsesWith(chain->result_id(), replacementId);
+    } else {
+      // Replace with a use of the variable.
+      context()->ReplaceAllUsesWith(chain->result_id(), var->result_id());
+    }
+  }
+
+  return true;
+}
+
+void ScalarReplacementPass::CreateReplacementVariables(
+    ir::Instruction* inst, std::vector<ir::Instruction*>* replacements) {
+  ir::Instruction* type = GetStorageType(inst);
+  uint32_t elem = 0;
+  switch (type->opcode()) {
+    case SpvOpTypeStruct:
+      type->ForEachInOperand([this, inst, &elem, replacements](uint32_t* id) {
+        CreateVariable(*id, inst, elem++, replacements);
+      });
+      break;
+    case SpvOpTypeArray:
+      for (uint32_t i = 0; i != GetArrayLength(type); ++i) {
+        CreateVariable(type->GetSingleWordInOperand(0u), inst, i, replacements);
+      }
+      break;
+
+    case SpvOpTypeMatrix:
+    case SpvOpTypeVector:
+      for (uint32_t i = 0; i != GetNumElements(type); ++i) {
+        CreateVariable(type->GetSingleWordInOperand(0u), inst, i, replacements);
+      }
+      break;
+
+    default:
+      assert(false && "Unexpected type.");
+      break;
+  }
+}
+
+void ScalarReplacementPass::CreateVariable(
+    uint32_t typeId, ir::Instruction* varInst, uint32_t index,
+    std::vector<ir::Instruction*>* replacements) {
```

Comment:
Somewhere in here you might need to copy decorations from the original struct to the new variable.

Reply:
I'm not sure that's appropriate, but I will peruse the decorations section in the spec.

Reply:
I've added code to handle decorations.